### DISCUSSION TITLE
Configure custom dap events?

### DISCUSSION THREAD

**Discussion by frere-jacques** [mfussenegger/nvim-dap#1259] [General]
I started configuring probe-rs for nvim-dap. This is a tool for embedded rust development.

I managed to get it running. But there is one feature I couldn't make use of yet and I don't know whether it's achievable at the moment.

probe-rs allows to configure rtt to provide some debug messages, prints etc. Since rtt is not available in dap, they use dap events to make the messages available over dap for their vscode plugin.

I had a look at the docs for nvim-dap and the event listener. To my understanding those are predefined events. Would it be possible to listen to other events and print them to the dap output or spawn an output terminal?

I looked in the dap.log and found the following:
```
[ WARN ] 2024-06-08T11:41:03Z+0200 ] ~/.local/share/nvim/lazy/nvim-dap/lua/dap/session.lua:1016 ]	"No event handler for "	{
  body = {
    channelName = "defmt",
    channelNumber = 0,
    dataFormat = "Defmt"
  },
  event = "probe-rs-rtt-channel-config",
  seq = 197,
  type = "event"
}
```
the dap.listerners contain
```                                              
    ["event_probe-rs-rtt-channel-config"] = {},
```
in before and after.

Edit:
I managed to listen to the events and write out contents of the events. Writing (or logging?) them into the repl terminal would be nicer.

The biggest issue is that the event "event_probe-rs-rtt-channel-config" expects a dap response. If no response is send the debugger will not sent consecutive events for the rtt data. I opened a [feature request](https://github.com/mfussenegger/nvim-dap/issues/1261) to create dap respnses from listeners.

**Comment by mfussenegger** [comment]
> I had a look at the docs for nvim-dap and the event listener. To my understanding those are predefined events. Would it be possible to listen to other events

It's possible to listen to any `event` send from a debug adapter, even if they are not part of the official specification.

>     ["event_probe-rs-rtt-channel-config"] = {},


Make sure to add the plugin-id key, and it must be a function:


```lua
local dap = require("dap")
dap.listeners.before["event_probe-rs-rtt-channel-config"]["my-plugin"] = function(session, body)
end
```

**Reply by frere-jacques** [reply]
Is "my-plugin" the id in dap.adapters?
I guess, in your snippet I have to define how to handle the events between line 2 and 3.

Sorry, I am really a noob on lua and configuring nvim stuff.

BTW: Are you interested in the config to add it in the wiki?

**Reply by frere-jacques** [reply]
And maybe again a noob question. If I manage to set up the handler, will I see the events in the repl or will they be placed to dap.log?

**Reply by frere-jacques** [reply]
If I look in your personal dotfiles I guess now the "my-plugin" refers to the path of the lua module

**Reply by frere-jacques** [reply]
@mfussenegger I tried to have a look in your dotfiles and the docs to understand how that works, but I think I am still missing a bit.
I modified my lua module in ~/.config/nvim/lua/plugins/nvim-dap-probe-rs.lua to the following:
```lua
return {
  "mfussenegger/nvim-dap",
  opts = function()
    local dap = require "dap"
    if not dap.adapters then dap.adapters = {} end
    dap.adapters["probe-rs-debug"] = {
      type = "server",
      port = "${port}",
      executable = {
        command = vim.fn.expand "$HOME/.cargo/bin/probe-rs",
        args = { "dap-server", "--port", "${port}" },
      },
    }
    require("dap.ext.vscode").type_to_filetypes["probe-rs-debug"] = { "rust" }
    dap.listeners.before["event_probe-rs-rtt-channel-config"]["plugins.nvim-dap-probe-rs"] = function(session, body)
      print("it works before", vim.inspect(body))
    end
    dap.listeners.after["event_probe-rs-rtt-channel-config"]["plugins.nvim-dap-probe-rs"] = function(session, body)
      print("it works after", vim.inspect(body))
    end
  end,
}
```
Starting the debug session I now see in the dap.listeners eg.:
```
["event_probe-rs-rtt-channel-config"] = {                                                 
      ["plugins.nvim-dap-probe-rs"] = <function 14>                                           
    },        
```

So far that stops the warning in the .cache/nvim/dap.log. But I don't get a print in the repl window.
And I didn't really understood the split in before and after yet.

Edit:
I added the following:
```lua
dap.listeners.after["event_probe-rs-show-message"]["plugins.nvim-dap-probe-rs"] = function(session, body)
    print("it works after", vim.inspect(body.message))
end
```
When I pull out my st debugger probe, probe-rs sends an event with errors. I saw that I get the print as a message in the bottom of nvim, where outputs of commands or request for user feedback is shown. But I rather want to see it in the repl.
But it is a step forward.

Edit2: I found out that if I have the function body blank, nothing happens. If I have only the vim.inspect in it nothing happens. If I have the print, I see the print and have to confirm it via pressing a key. But if I toggle back to nvim-dap-ui (it closes, because the session gets killed, when I pull the USB dongle), I see the messages and a lot more:
```
[debug-adapter stderr]  WARN probe_rs::session:788: Could not clear all hardware breakpoints: An ARM specific error occurred.

Caused by:
    0: The debug probe encountered an error.
    1: An error which is specific to the debug probe in use occurred.
    2: USB
 WARN probe_rs::session:798: Failed to deconfigure device during shutdown: An ARM specific error occurred.

Caused by:
    0: The debug probe encountered an error.
    1: An error which is specific to the debug probe in use occurred.
    2: USB
probe-rs-debug: Session ended: An ARM specific error occurred.
```
This has to be connected to my listener config, because if I comment the function out and repeat, I get nothing there. Is it possible to get those messages without printing with confirm in the bottom of nvim?

**Reply by frere-jacques** [reply]
The print() is not really a good solution. It vanishes sometimes fastly. I set the functions now up to write to a file. That is probably handy, but a direct print inside nvim-dap-ui would be nice. But I don't found how to write (or log?) there.

**Reply by Tzlzl12** [reply]
maybe you can use vim.notify(message, level, { timeout = the number you want(false is last forever, if i dont misremember)})
A internative of print

**Comment by frere-jacques** [comment]
I think I managed to get it as far as I could. There might be nicer solutions to get it enabled, and potential stuff like a down channel is not easily achievable, but so far it works. Will provide a config for the wiki soon.
